src/bindfs.c

```diff
@@ -769,6 +771,234 @@ static int bindfs_fgetattr(const char *path, struct stat *stbuf,
+        DPRINTF("Could not lstat '%s': %s", real_path, strerror(errno));
+        free(real_path);
+        return -errno;
+    }
+
+    uid_t euid = geteuid();
+    gid_t egid = getegid();
+
+    // TODO: There are `root` cases still needing to be accounted for.
+    //       This is not the correct implementation!
+    //if (euid == 0)
+    //    return 0;
+
+    struct passwd *pwd = getpwuid(euid);
+
+    // We can perform this check early, Unsubstantiated, but I would imagine
+    // it's one of the more likely cases, so it makes sense to optimise for it.
+    //
+    // If the user ID of the process is the owner, the owner entry determines
+    // access.
+    if (euid == st.st_uid) {
+        return (((st.st_mode & 0700) >> 6) & wants) == wants
+            ? 0
+            : -EACCES;
+    }
+
+    // Get the groups the effective user is a part of
+    int ngroups = 8;
+    gid_t *groups = malloc(sizeof(*groups) * ngroups);
+
+    while (getgrouplist(pwd->pw_name, egid, groups, &ngroups) == -1) {
+        DPRINTF("Reallocating space for groups to %d entries", ngroups);
+        groups = realloc(groups, sizeof(*groups) * ngroups);
+    }
+
+    // Notes on the below variables:
+    // We do not need to store the owner permissions as we can exit immediately
+    // if we find them and do not need to do further processing.
+
+    // We do not need to store group ACL permissions separately from the
+    // standard group permissions. We can merge all the group permissions
+    // together and treat them as one. TODO: I think, need to thoroughly test.
+    // NOTE: This only holds if an object that has ACLs where `group_a = r--`,
+    // `group_b = --x` and the request is for `r-x` (`wants` = 5) we should
+    // grant access, if we should disallow access, we can't aggregate
+    // permissions in this way.
+
+    unsigned int group_perms = 0;
+    unsigned int other_perms = 0;
+
+    unsigned int user_acl_perms = 0;
+    unsigned int acl_mask = 0;
+
+    unsigned int has_user_acl = 0;
+    unsigned int has_group = 0;
+    unsigned int has_mask = 0;
+
+    acl_t acl = acl_get_file(real_path, ACL_TYPE_ACCESS);
+    acl_entry_t acl_entry;
+    for (
+        int result = acl_get_entry(acl, ACL_FIRST_ENTRY, &acl_entry);
+        result > 0;
+        result = acl_get_entry(acl, ACL_NEXT_ENTRY, &acl_entry)
+    ) {
+        acl_tag_t acl_tag;
+        if (acl_get_tag_type(acl_entry, &acl_tag) != 0) {
+            DPRINTF("Could not get ACL tag type: %s", strerror(errno));
+            continue;
+        }
+
+        acl_permset_t acl_permset;
+        if (acl_get_permset(acl_entry, &acl_permset) != 0) {
+            DPRINTF("Could not get ACL permset: %s", strerror(errno));
+            continue;
+        }
+
+        unsigned int permset_as_bits = permset_to_bits(acl_permset);
+
+        void *acl_qualifier = acl_get_qualifier(acl_entry);
+        switch(acl_tag) {
+            case ACL_USER_OBJ:
+                if (euid == st.st_uid) {
```

Comment:
You are correct. We should be able to do away with this, I'm probably overly cautious, but the clause on line 840 is going to hit on the standard Unix permission and this will hit on the ACL version of the same permission. They should be the same, but for the sake of a few extra bytes, I thought I'd include it. Your call, I can clarify this in a comment or remove it?

Reply:
Just looking naively at the code, since nothing changes `euid` nor `st.st_uid` after the previous check, this `if` can never actually be entered.

So it seems either this condition is written incorrectly, or we can just do nothing here (with a comment explaining why), or perhaps doing the check early is incorrect after all. I've not looked at ACLs well enough yet to know which it is.

Reply:
You are correct. I should have been more specific in my previous comment. Technically they're checking two different things, but the outcome is the same if we remove the latter check. In my mind it was partly for documenting the full algorithm and partly for robustness if things were to be changed in the future without full consideration. I'm happy to remove this check and write a comment.